applications/snake-game/snake-game.c

```diff
@@ -0,0 +1,419 @@
+    snake_state->points[0] = next_step;
+}
+
+static void snake_game_process_game_step(SnakeState* const snake_state) {
+    if(snake_state->state == GameStateGameOver) {
+        return;
+    }
+
+    bool can_turn = (snake_state->points[0].x % 2 == 0) && (snake_state->points[0].y % 2 == 0);
+    if(can_turn) {
+        snake_state->currentMovement = snake_game_get_turn_snake(snake_state);
+    }
+
+    Point next_step = snake_game_get_next_step(snake_state);
+
+    bool crush = snake_game_collision_with_frame(next_step);
+    if(crush) {
+        if(snake_state->state == GameStateLife) {
+            snake_state->state = GameStateLastChance;
+            return;
+        } else if(snake_state->state == GameStateLastChance) {
+            snake_state->state = GameStateGameOver;
+            return;
+        }
+    } else {
+        if(snake_state->state == GameStateLastChance) {
+            snake_state->state = GameStateLife;
+        }
+    }
+
+    crush = snake_game_collision_with_tail(snake_state, next_step);
+    if(crush) {
+        snake_state->state = GameStateGameOver;
+        return;
+    }
+
+    bool eatFruit = (next_step.x == snake_state->fruit.x) && (next_step.y == snake_state->fruit.y);
+    if(eatFruit) {
+        snake_state->len++;
+        if(snake_state->len >= MAX_SNAKE_LEN) {
+            snake_state->state = GameStateGameOver;
+            return;
+        }
+    }
+
+    snake_game_move_snake(snake_state, next_step);
+
+    if(eatFruit) {
+        snake_state->fruit = snake_game_get_new_fruit(snake_state);
+    }
+}
+
+int32_t snake_game_app(void* p) {
+    srand(DWT->CYCCNT);
+
+    osMessageQueueId_t event_queue = osMessageQueueNew(8, sizeof(SnakeEvent), NULL);
+
+    SnakeState* snake_state = furi_alloc(sizeof(SnakeState));
+    snake_game_init_game(snake_state);
+
+    ValueMutex state_mutex;
+    if(!init_mutex(&state_mutex, snake_state, sizeof(SnakeState))) {
+        furi_log_print(FURI_LOG_ERROR, "cannot create mutex\r\n");
+        free(snake_state);
+        return 255;
+    }
+
+    ViewPort* view_port = view_port_alloc();
+    view_port_draw_callback_set(view_port, snake_game_render_callback, &state_mutex);
+    view_port_input_callback_set(view_port, snake_game_input_callback, event_queue);
+
+    osTimerId_t timer =
+        osTimerNew(snake_game_update_timer_callback, osTimerPeriodic, event_queue, NULL);
+    osTimerStart(timer, osKernelGetTickFreq() / 4);
+
+    // Open GUI and register view_port
+    Gui* gui = furi_record_open("gui");
+    gui_add_view_port(gui, view_port, GuiLayerFullscreen);
+
+    SnakeEvent event;
+    for(bool processing = true; processing;) {
+        osStatus_t event_status = osMessageQueueGet(event_queue, &event, NULL, 100);
+
+        SnakeState* snake_state = (SnakeState*)acquire_mutex_block(&state_mutex);
+
+        if(event_status == osOK) {
+            // press events
+            if(event.type == EventTypeKey) {
+                if(event.input.type == InputTypeShort) {
```

Comment:
`InputTypeShort` is high level event that is generated upon button release(which introduces double debounce delay).

That makes control sluggish and unpredictable. Use low level event `InputTypePress` instead. 

Also fruit placer becomes quite predictable when you moving in horizontal plane. But it is not critical.